### PR TITLE
docs: improve documentation for audit release

### PR DESCRIPTION
**Pull request by jaybxyz**
## Description

This PR is the last one to improve documentation quality for the upcoming audit release before we work on #172.
Let's make documentation easy to read and resourceful. Make sure that all the necessary contents are there for other developers and auditors to test.

## Tasks

- [x] Update README.md (make it easy and intuitive for other developers to quickly get on board with farming module)
- [x] Update TECHNICAL-SETUP.md
- [x] Update `spec` docs up to date

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: x/farming/spec/01_concepts.md [hunk 1]
```diff
@@ -27,3 +27,16 @@ A `FixedAmountPlan` distributes fixed amount of coins to farmers for every epoch
...
+- AUR for each staking coin for each epoch can be calculated as below
+  - ![](https://latex.codecogs.com/svg.latex?\Large&space;\sum_{i=0}^{now}\frac{TR_i}{TS_i})
+    - ![](https://latex.codecogs.com/svg.latex?\Large&space;i) : each `EpochId`
+    - ![](https://latex.codecogs.com/svg.latex?\Large&space;now) : current `EpochId`
+    - ![](https://latex.codecogs.com/svg.latex?\Large&space;TS_i) : total staking amount of the staking coin for epoch i
+    - ![](https://latex.codecogs.com/svg.latex?\Large&space;TR_i) : total reward amount of the staking coin for epoch i
+- Accumulated rewards from any staking position can be calculated from AUR and the staking amount of the position as below
+  - ![](https://latex.codecogs.com/svg.latex?\Large&space;x*\(\sum_{i=0}^{now}\frac{TR_i}{TS_i}-\sum_{i=0}^{start}\frac{TR_i}{TS_i}\))
+    - assuming constant staking amount for the staking epochs
+    - ![](https://latex.codecogs.com/svg.latex?\Large&space;x) : staking amount for the staking period
```
Comment (hallazzang) [inline x/farming/spec/01_concepts.md:42]: We use the name `HistoricalRewards` which is similar to `ValidatorHistoricalRewards` in `x/distribution`. We need to update this paragraph to match with the code.

File: x/farming/spec/05_end_block.md [hunk 2]
```diff
@@ -1,15 +1,6 @@
...
-        - keep stake, reward states for unstakable stakes and claimable rewards each farmers
-        - rest of the fund in `farmingPoolAddress` sent to `terminationAddress`
\ No newline at end of file
+At the end of each block, it terminates plans that their end time has passed over the current block time. It sends all remaining coins in the plan's farming pool account `FarmingPoolAddress` to the termination address `TerminationAddress` and mark the plan as terminated by making `Terminated` false. A global parameter `NextEpochDays` is there but the farming module uses an internal state `CurrentEpochDays` to prevent from a case where it may affect the rewards allocation. Suppose `NextEpochDays` is 7 and it is proposed to change the value to 1 through governance proposal. Although the proposal is passed, rewards allocation should continue to proceed with 7, not 1. The [test code](https://github.com/tendermint/farming/blob/69db071ce3/x/farming/abci_test.go#L12-L64) is available for you if you want to understand it in more detail. Then it allocates farming rewards, processes `QueueCoins` to `StakedCoins`, and sets `LastEpochTime` to keep in track in case of the chain upgrdae.
```
Comment (hallazzang) [inline x/farming/spec/05_end_block.md:6]: It sends all remaining coins in the plan's farming pool account `FarmingPoolAddress` to the termination address `TerminationAddress` and mark the plan as terminated by making `Terminated` ~false~ true.

Then it allocates farming rewards, processes ~`QueueCoins`~ `QueuedStaking` to ~`StakedCoins`~ be staked, and sets `LastEpochTime` to keep in track in case of the chain ~upgrdae~ upgrade.

**Review by hallazzang — CHANGES_REQUESTED** [review]

File: x/farming/spec/05_end_block.md (hunk 2 above)
Comment (kogisin) [reply x/farming/spec/05_end_block.md:6]: Great! catch! I'll make the changes.

File: x/farming/spec/03_state_transitions.md [hunk 3]
```diff
@@ -25,52 +25,57 @@ const (
...
-    - assuming constant staking amount for the staking epochs
-    - ![](https://latex.codecogs.com/svg.latex?\Large&space;x) : staking amount for the staking period
\ No newline at end of file
+## Stake
+
+When a farmer stakes an amount of coins, the following state transitions occur:
+
+- it reserves the amount of coins to the staking reserve pool account `StakingReservePoolAcc` 
+- it creates `QueuedStaking` object and stores the staking coins in `QueueStaking`, which are waiting in a queue until the end of epoch to move to `Staking` object
+- it imposes more gas if the farmer already has `Staking` with the same coin denom
```
Comment (dongsam) [inline x/farming/spec/03_state_transitions.md:52]: It would be modified by https://github.com/tendermint/farming/pull/175/files#r730642438 @hallazzang

File: x/farming/spec/05_end_block.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kogisin) [reply x/farming/spec/05_end_block.md]: I'll close this since it is already fixed. Thanks!

File: x/farming/spec/01_concepts.md [hunk 4]
```diff
@@ -27,3 +27,26 @@ A `FixedAmountPlan` distributes fixed amount of coins to farmers for every epoch
 
 A `RatioPlan` distributes to farmers by ratio distribution for every epoch day. If the plan creators `FarmingPoolAddress` is depleted with distributing coins, then there is no more coins to distribute unless it is filled up with more coins.
 
+## Accumulated Reward Calculation
+
+In farming module, farming rewards are calculated per epoch based on plans. The rewards for a single farmer can be calculated by taking the total rewards for the epoch before the staking started, minus the current total rewards. The farming module takes references from [F1 Fee Distribution](https://github.com/cosmos/cosmos-sdk/blob/master/docs/spec/fee_distribution/f1_fee_distr.pdf) that is used in Cosmos SDK [x/distribution](https://github.com/cosmos/cosmos-sdk/blob/master/x/distribution/spec/01_concepts.md) module.
+
+### Accumulated Unit Reward 
+
+AUR represents `HistoricalRewards` that are accumulated rewards for each staking coin with amount 1.
```
Comment (hallazzang) [inline x/farming/spec/01_concepts.md:36]: I suggest:
> `HistoricalRewards` represents accumulated rewards for each staking coin with unit amount 1.

File: x/farming/spec/01_concepts.md [hunk 5]
```diff
@@ -27,3 +27,26 @@ A `FixedAmountPlan` distributes fixed amount of coins to farmers for every epoch
...
+
+### Base Algorithm 
+
+`HistoricalRewards` for each staking coin for every epoch can be calculated as the following algorithm:
+
+- ![](https://latex.codecogs.com/svg.latex?\Large&space;\sum_{i=0}^{now}\frac{TR_i}{TS_i})
+    - ![](https://latex.codecogs.com/svg.latex?\Large&space;i) : `StartingEpoch`
```
Comment (hallazzang) [inline x/farming/spec/01_concepts.md:43]: `i` is not `StartingEpoch`, it's just **each epoch**.

File: x/farming/spec/03_state_transitions.md [hunk 6]
```diff
@@ -22,55 +22,39 @@ const (
...
+- it subtracts the unstaking amount of coins from `QueueStaking` first and if it is not sufficient then it subtracts from `Staking`
+- it releases the unstaking amount of coins to the farmer
+
+## Harvest (Reward Withdrawal)
+
+- it calculates `CumulativeUnitRewards` in `HistoricalRewards` object in order to get the rewards for the staking coin denom which are accumulated over the last epochs for the farmer
+- it releases the accumulated rewards to the farmer if it is not zero and decreases the `OutstandingRewards`
+- it sets `StartingEpoch` in `Staking` object
+
+## Reward Allocation
```
Comment (hallazzang) [inline x/farming/spec/03_state_transitions.md:48]: I think it'd be good to add this sentence:
> If the sum of total calculated `EpochAmount`(or `EpochRatio` multiplied by the farming pool's balance) exceeds the farming pool's balance, then the reward allocation is skipped for that epoch.

File: x/farming/spec/03_state_transitions.md [hunk 7]
```diff
@@ -22,55 +22,39 @@ const (
...
+
+Each abci end block call, the operations to update rewards allocation are to execute:
+
+++ https://github.com/tendermint/farming/blob/69db071ce30b99617b8ba9bb6efac76e74cd100b/x/farming/keeper/reward.go#L363-L426
+
+- it calculates rewards allocation information for the end of the current epoch depending on plan type `FixedAmountPlan` or `RatioPlan`
+- it distributes total allocated coins from each plan’s farming pool address `FarmingPoolAddress` to the rewards reserve pool account `RewardsReserveAcc`
+- it calculates staking coin weight for each denom in each plan and gets the unit rewards by denom
+- it updates `HistoricalRewards` and `CurrentEpoch` based on the allocation information
+- it automatically withdraws the accumulated rewards to the farmer with the given staking coin denom if `Staking` position exists
```
Comment (hallazzang) [inline x/farming/spec/03_state_transitions.md:58]: I think we should remove this line, since it's not related with the reward allocation. Instead, it's related with `Stake` action.

File: x/farming/spec/03_state_transitions.md [hunk 8]
```diff
@@ -22,55 +22,39 @@ const (
...
+- it deletes `QueueStaking` object after moving `QueueCoins` to `StakedCoins` in `Staking` object
+- it increases `TotalStakings` for the staking coin denom
```
Comment (hallazzang) [inline x/farming/spec/03_state_transitions.md:60]: Also, these are not related with reward allocation, instead, these are related with `ProcessQueuedCoins` method.
I think we need a separate section for these end of epoch events.

File: x/farming/spec/03_state_transitions.md (hunk 6 above)
Comment (kogisin) [reply x/farming/spec/03_state_transitions.md:48]: Let me add the line in the beginning of "Reward Allocation" section.

File: x/farming/spec/03_state_transitions.md (hunk 7 above)
Comment (kogisin) [reply x/farming/spec/03_state_transitions.md:58]: As I asked @hallazzang offline, he refers to line 58 only.
